File: 인터넷 강의/python_ruby/Inheritance/2.py
class Cal(object):
    # 생성자의 첫번째 매개변수에 self를 놓아야한다
    # ! 생성자 keyword는 __init__이다.
    def __init__(self, v1, v2):
        self.v1 = v1
        self.v2 = v2

    # 메소드에 매개변수 self반드시 필요하다.
    def add(self):
        return self.v1 + self.v2

    def subtract(self):
        return self.v1 - self.v2

    def setV1(self, v):
        # 내장되어있는 instance인지 검사
        if isinstance(v, int):
            self.v1 = v
    
    def getV1(self):
        return self.v1


class CalMultiply(Cal):
    def multiply(self):
        return self.v1 * self.v2

c1 = CalMultiply(10, 10)
print(c1.multiply())

File: 인터넷 강의/python_ruby/Override/2.rb
class Cal
    @@_history = []

    def initialize(v1, v2)
        @v1 = v1
        @v2 = v2
    end
    #! @가 붙으면 인스턴스 변수이다.

    def add()
        result = @v1 + @v2
        @@_history.push("#{@v1}  + #{@v2} = #{result}")
        return result
    end

    def subtract()
        result = @v1 - @v2
        @@_history.push("#{@v1}  - #{@v2} = #{result}")
        return result
    end

    def setV1(v)
        if v.is_a?(Integer)
            @v1 = v
        end
    end

    def getV1()
        return @v1
    end

    def Cal.history()
        for item in @@_history
            p item
        end
    end

    def info()
        return "Cal => v1 : #{@v1}, v2 : #{@v2} "
    end
end

class CalMultiply < Cal
    def multiply()
        result = @v1 * @v2
        @@_history.push("#{@v1}  * #{@v2} = #{result}")
        return result
    end

    def info()
        return "CalMultiply => #{super()}"
    end
end

class CalDivide < CalMultiply
    def divide()
        result = @v1 / @v2
        @@_history.push("#{@v1}  / #{@v2} = #{result}")
        return result
    end

    def info()
        return "CalMultiply => #{super()}"
    end
end

c0 = Cal.new(30, 60)
p c0.info()
c1  = CalMultiply.new(10, 10)
p c1.info()
c2 = CalDivide.new(20, 10)
p c2.info()